### PR TITLE
Add search bar component

### LINKED ISSUE #79: [Feature]: Add Search Bar component
### To which use case does your feature request relate?

_No response_

### Is your feature request related to a problem? Please describe

Currently, we don't have any custom component for a search bar (used for searching shortcuts, papers, references, ...).

### Describe the solution you'd like

Add a component for the Search Bar.

### Describe alternatives you've considered

Alternatively, we could add a style for a shadcn input component.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Merseleo**
Closes #79

## What I have made

- I added a search bar component, which allows users to search, whereas the search function is passed to the component
- I updated the *vite.config.ts* to ignore the routes and app.d.ts from coverage (tested by e2e tests)

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (it is a component which has no units to test)
  - [x] integration tests
  - ~~[ ] end-to-end tests~~ (it is a component which can not be tested with e2e tests)
- [x] (for reviewer) I have checked the implementation against the requirements

### CODE REVIEW

File: src/lib/components/composites/search-bar/SearchBar.svelte [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    interface Props {
+        placeholderText?: string;
+        onSearch: (searchText: string) => void;
+    }
+    const { placeholderText = "Search", onSearch }: Props = $props();
+
+    let searchInput: string = $state("");
+
+    // timeout in ms to wait after last key input before start search
+    const timeout: number = 500;
```
Comment (Slartibartfass2) [inline src/lib/components/composites/search-bar/SearchBar.svelte]: If the unit is not inferred by type, I would add it in the name.
```suggestion
    const timeoutInMs: number = 500;
```

File: src/lib/components/composites/search-bar/SearchBar.svelte [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            if (timeoutId !== null) clearTimeout(timeoutId);
+
+            // Wait a certain amount of time and start search, if no new key is pressed
+            timeoutId = setTimeout(() => onSearch(searchInput), timeout);
+        }
+    };
+</script>
+
+<!--
+@component
```
Comment (Slartibartfass2) [inline src/lib/components/composites/search-bar/SearchBar.svelte:44]: Great that you can add documentation for components like this. Didn't know that.

File: src/lib/components/composites/search-bar/SearchBar.svelte [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+It starts the search on the following events:
+  - the user pressed enter
+  - the user click the search icon
+  - the user enters the search text and wait
+If the user enters the 'Esc' key, the search is aborted.
+
+Usage:
+```tsx
+    <SearchBar placeholderText={yourText} onSearch={yourCallback    } />
```
Comment (Slartibartfass2) [inline src/lib/components/composites/search-bar/SearchBar.svelte]: Let's keep the description in one tense and use the svelte highlighting (tsx is for react + typescript).
```suggestion
It starts the search on the following events:
  - the user presses enter
  - the user clicks the search icon
  - the user enters the search text and waits
If the user enters the 'Esc' key, the search is aborted.

Usage:
```svelte
    <SearchBar placeholderText={yourText} onSearch={yourCallback} />
```

File: src/routes/+page.svelte [hunk 4]
```diff
@@ -1,10 +1,20 @@
...
 </script>
 
-<SimpleNavigationBar {user} title="SnowballR" tabs={[]} defaultTabValue="" />
+<div class="flex flex-col gap-5">
+    <!-- example ussage of the navigation bar -->
+    <SimpleNavigationBar {user} title="SnowballR" tabs={[]} defaultTabValue="" />
+
+    <!-- example usage of the search bar ... -->
+    <SearchBar placeholderText={searchPlaceholderText} onSearch={(input) => console.log(input)} />
+    <SearchBar onSearch={(input) => console.log(input)} />
```
Comment (Slartibartfass2) [inline src/routes/+page.svelte]: This is no example usage for the navigation bar, it is, it's actual usage.
Let's move the examples to the theme page or maybe another page called examples, to keep the demos but don't block the pages from being designed.

File: src/lib/components/primitives/input/input.svelte [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        value = $bindable(),
+        class: className,
+        ...restProps
+    }: WithElementRef<HTMLInputAttributes> = $props();
+</script>
+
+<input
+    bind:this={ref}
+    class={cn(
+        "border-input bg-background ring-offset-background placeholder:text-muted-foreground flex h-10 w-full rounded-md border px-3 py-2 text-default file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50",
```
Comment (Slartibartfass2) [inline src/lib/components/primitives/input/input.svelte]: Let's use the theme that we already defined.
```suggestion
        "border-input-border-slate bg-background ring-offset-background placeholder:text-placeholder flex h-fit w-full rounded-md border px-4 py-2 text-default file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50",
```

I also noticed that the `text-placeholder style doesn't use the predefined color for it, can you change that:
```pcss
.text-placeholder {
    @apply text-base font-medium tracking-normal text-input-placeholder-slate italic;
}
```

File: src/lib/components/composites/search-bar/SearchBar.svelte [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+```
+-->
+<div class="relative w-full">
+    <!-- inspired by https://github.com/shadcn-ui/ui/issues/1562 -->
+    <Input
+        type="text"
+        placeholder={placeholderText}
+        bind:value={searchInput}
+        onkeyup={processNewInput}
+        class="px-4 py-3 pr-10"
```
Comment (Slartibartfass2) [inline src/lib/components/composites/search-bar/SearchBar.svelte]: Now that this style is already in input, you only have to make space for the search button.
```suggestion
        class="pr-10"
```

File: test/integration/search-bar/search-bar.test.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import { render, screen } from "@testing-library/svelte";
+import userEvent from "@testing-library/user-event";
+
+describe("SearchBarComponent", () => {
+    test("When all required props are provided, then the search bar does not throw any error", () => {
+        render(SearchBar, {
+            props: { onSearch: (input: string) => console.log(input) },
+        });
+        assert.doesNotThrow(() => screen.getByPlaceholderText("Search"));
+        expect(screen.getByPlaceholderText("Search")).toBeInTheDocument();
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]: These assertions do the same things, don't they?

File: test/integration/search-bar/search-bar.test.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    });
+
+    test("When placeholder text prop is provided, then the search bar show the provided placeholder text and not the default text", () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]:
```suggestion
    test("When placeholder text prop is provided, then the search bar shows the provided placeholder text and not the default text", () => {
```

File: test/integration/search-bar/search-bar.test.ts
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts:1]: Let's keep the When ..., then ... style consistent in this file.

File: test/integration/search-bar/search-bar.test.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                placeholderText: "Search paper",
+                onSearch: (input: string) => console.log(input),
+            },
+        });
+
+        assert.throws(() => screen.getByPlaceholderText("Search"));
+        assert.doesNotThrow(() => screen.getByPlaceholderText("Search paper"));
+    });
+
+    test("When the user type a search string into the search bar and click the search icon, the search is conducted", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]:
```suggestion
    test("When the user enters a search string into the search bar and click the search icon, then the search is conducted", async () => {
```

File: test/integration/search-bar/search-bar.test.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        render(SearchBar, {
+            props: { onSearch: (input: string) => (searchInput = input) },
+        });
+
+        await userEvent.type(screen.getByRole("textbox"), "Test");
+        await userEvent.click(screen.getByRole("button"));
+        assert.equal(searchInput, "Test");
+    });
+
+    test("When the user type a search string into the search bar and press enter, the search is conducted", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]:
```suggestion
    test("When the user enters a search string into the search bar and press enter, then the search is conducted", async () => {
```

File: test/integration/search-bar/search-bar.test.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        render(SearchBar, {
+            props: { onSearch: (input: string) => (searchInput = input) },
+        });
+
+        await userEvent.type(screen.getByRole("textbox"), "Test");
+        await userEvent.type(screen.getByRole("textbox"), "{enter}");
+        assert.equal(searchInput, "Test");
+    });
+
+    test("When the user type a search string into the search bar waits, the search is conducted", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]:
```suggestion
    test("When the user enters a search string into the search bar waits, then the search is conducted", async () => {
```

File: test/integration/search-bar/search-bar.test.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+        await userEvent.type(screen.getByRole("textbox"), "Test");
+        assert.equal(searchInput, "start");
+
+        setTimeout(() => {
+            assert.equal(searchInput, "Test");
+        }, 500);
+    });
+
+    test("When the user type no search string into the search bar, no search is conducted", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]:
```suggestion
    test("When the user doesn't enter a search string into the search bar, then no search is conducted", async () => {
```

File: test/integration/search-bar/search-bar.test.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        });
+
+        await userEvent.type(screen.getByRole("textbox"), "{enter}");
+        assert.equal(searchInput, "start");
+
+        await userEvent.click(screen.getByRole("button"));
+        assert.equal(searchInput, "start");
+    });
+
+    test("When the user escapes the search bar, no search is conducted and the search input cleared", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]:
```suggestion
    test("When the user escapes the search bar, then no search is conducted and the search input cleared", async () => {
```

File: test/integration/search-bar/search-bar.test.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        let searchInput: string = "start";
+        render(SearchBar, {
+            props: { onSearch: (input: string) => (searchInput = input) },
+        });
+
+        await userEvent.type(screen.getByRole("textbox"), "Test");
+        await userEvent.type(screen.getByRole("textbox"), "{esc}");
+        assert.equal(searchInput, "start");
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]: According to coverage the case for escape isn't executed. Is this the correct format to simulate key events?

**Review by Slartibartfass2 — CHANGES_REQUESTED** [review]
Great job, the search bar looks great and the logic looks reasonable. The documentation is as good as expected :)

Just some minor changes left.

I noticed that the `onSearch` method is triggered again, when the user navigates in the text with the arrow buttons. Maybe add a check whether the input has changed and only then trigger the method, except when the enter key is pressed again.

**Comment by Slartibartfass2** [comment]
Minor side quest:
I would add `exclude: ["src/routes/**/+*.{svelte,ts}", "src/app.d.ts"],` to the `coverage` object in `vite.config.ts` to only include components and logic into our coverage reports. The pages are later tested by the e2e tests.

Somehow the branch coverage isn't shown, but this needs to be solved separately.

**Comment by Slartibartfass2** [comment]
[This](https://testing-library.com/docs/svelte-testing-library/example#two-way-data-binding) could also be interesting for testing the search input.

**Comment by Merseleo** [comment]
> Great job, the search bar looks great and the logic looks reasonable. The documentation is as good as expected :)
> 
> Just some minor changes left.
> 
> I noticed that the `onSearch` method is triggered again, when the user navigates in the text with the arrow buttons. Maybe add a check whether the input has changed and only then trigger the method, except when the enter key is pressed again.

Great catch, thank you. I fixed it.

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 1 above)
Comment (Merseleo) [reply src/lib/components/composites/search-bar/SearchBar.svelte]: Agree, I will change it (and keep it in mind for similar cases).

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 2 above)
Comment (Merseleo) [reply src/lib/components/composites/search-bar/SearchBar.svelte:44]: Yep, I doesn't work in Webstorm, but I thought we should add it for all VS Code users.

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 3 above)
Comment (Merseleo) [reply src/lib/components/composites/search-bar/SearchBar.svelte]: I changed it.

File: src/routes/+page.svelte (hunk 4 above)
Comment (Merseleo) [reply src/routes/+page.svelte]: I create a separate page, where we can put our custom components, so the homepage doesn't contain them anymore.

File: src/lib/components/primitives/input/input.svelte (hunk 5 above)
Comment (Merseleo) [reply src/lib/components/primitives/input/input.svelte]: I adapted the .text-placeholder style as well as the input styling (matching the theme from Figma).

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 6 above)
Comment (Merseleo) [reply src/lib/components/composites/search-bar/SearchBar.svelte]: Changed.

File: test/integration/search-bar/search-bar.test.ts
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts:1]: I will try it ...

File: test/integration/search-bar/search-bar.test.ts (hunk 7 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Good catch, I will delete one.

File: test/integration/search-bar/search-bar.test.ts (hunk 8 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Changed.

File: test/integration/search-bar/search-bar.test.ts (hunk 9 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Done.

File: test/integration/search-bar/search-bar.test.ts (hunk 10 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Done.

File: test/integration/search-bar/search-bar.test.ts (hunk 11 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Done.

File: test/integration/search-bar/search-bar.test.ts (hunk 12 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Done.

File: test/integration/search-bar/search-bar.test.ts (hunk 13 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Done.

File: test/integration/search-bar/search-bar.test.ts (hunk 14 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: I think there is indeed a problem, I will check it and try to solve it.

**Comment by Merseleo** [comment]
> Minor side quest: I would add `exclude: ["src/routes/**/+*.{svelte,ts}", "src/app.d.ts"],` to the `coverage` object in `vite.config.ts` to only include components and logic into our coverage reports. The pages are later tested by the e2e tests.
> 
> Somehow the branch coverage isn't shown, but this needs to be solved separately.

I added it.

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 1 above)
Comment (Slartibartfass2) [reply src/lib/components/composites/search-bar/SearchBar.svelte]: This change isn't applied yet.

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 3 above)
Comment (Slartibartfass2) [reply src/lib/components/composites/search-bar/SearchBar.svelte]: There's still the gigantic whitespace after `yourCallback`.

File: test/integration/search-bar/search-bar.test.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+                placeholderText: "Search paper",
+                onSearch: (input: string) => console.log(input),
+            },
+        });
+
+        assert.throws(() => screen.getByPlaceholderText("Search"));
+        assert.doesNotThrow(() => screen.getByPlaceholderText("Search paper"));
+    });
+
+    test("When the user enters a search string into the search bar and click the search icon, then the search is conducted", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]: Missed this last time.
```suggestion
    test("When the user enters a search string into the search bar and clicks the search icon, then the search is conducted", async () => {
```

File: test/integration/search-bar/search-bar.test.ts [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+            props: { onSearch: (input: string) => (searchInput = input) },
+        });
+
+        await userEvent.type(screen.getByRole("textbox"), "Test");
+        assert.notEqual(searchInput, "Test");
+        await userEvent.click(screen.getByRole("button"));
+        assert.equal(searchInput, "Test");
+    });
+
+    test("When the user enters a search string into the search bar and press enter, then the search is conducted", async () => {
```
Comment (Slartibartfass2) [inline test/integration/search-bar/search-bar.test.ts]: Missed this last time.
```suggestion
    test("When the user enters a search string into the search bar and presses enter, then the search is conducted", async () => {
```

**Review by Slartibartfass2 — CHANGES_REQUESTED** [review]
Nearly done!

I still have some questions regarding the behavior:
- Should we trigger the onSearch method again, when enter or the button is pressed if the input hasn't changed? Is this the refresh functionality?
- When there's input and I press escape, the onSearch method isn't triggered. I think this could be misleading, e.g. when filtering the papers in the project view I want to clear the filter when pressing escape

what's your opinion about those behaviors?

File: src/lib/components/composites/search-bar/SearchBar.svelte (hunk 1 above)
Comment (Merseleo) [reply src/lib/components/composites/search-bar/SearchBar.svelte]: Sorry, accidentally restored this change.

File: test/integration/search-bar/search-bar.test.ts (hunk 15 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: As i ...

File: test/integration/search-bar/search-bar.test.ts (hunk 16 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: Fixed it.

File: test/integration/search-bar/search-bar.test.ts (hunk 14 above)
Comment (Merseleo) [reply test/integration/search-bar/search-bar.test.ts]: The test should be corrected now.

**Review by Merseleo — COMMENTED** [review]
Regarding your questions on the behavior:
1) I would interpret it as a "refresh" option, I mean isn't this common search functionality that pressing the search button again or pressing enter multiple times start the search again and refreshes the results (at least DuckDuckGo etc. have this behavior).
2) Ok, I thought clearing the input with escape has the meaning of aborting the search. Same here: if you look at the behavior of search engines for the escape character, they often loose focus and don't search (and also not clear the input, but this is another discussion). But if you want, we can ask @luca-schlecker or @araschke what they think the behavior should be?!

**Review by Slartibartfass2 — APPROVED** [review]

**Comment by Slartibartfass2** [comment]
Merge after #102

**Pull request merged**